File: src/main/java/lab6/string_list/ListIterator.java
package lab6.string_list;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class ListIterator implements Iterator<String> {
    protected Node current;
    protected StringList list;

    ListIterator(Node current, StringList list) {
        this.current = current;
        this.list = list;
    }

    Node getCurrentNode() {
        return current;
    }

    String getCurrent() {
        return current == null ? null : current.getValue();
    }

    StringList getList() {
        return list;
    }

    @Override
    public boolean hasNext() {
        return current != null;
    }

    @Override
    public String next() {
        Node next = current;
        if (next == null) {
            throw new NoSuchElementException("Iterator doesn't point to next element!");
        }
        current = current.getNext();
        return next.getValue();
    }
}

File: src/main/java/lab7/my_list/ListIterator.kt
package lab7.my_list

open class ListIterator<T>(var current: Node<T>? = null, var array: MyList<T>) : Iterable<T>, Iterator<T> {
    fun getCurrentNode(): Node<T>? {
        return current
    }

    fun getCurrent(): T? {
        return if (current == null) null else current!!.getValue()
    }

    fun getList(): MyList<T>? {
        return array
    }

    override operator fun hasNext(): Boolean {
        return current != null
    }

    override operator fun next(): T {
        val next = current ?: throw NoSuchElementException("Iterator doesn't point to next element!")
        current = current!!.getNext()
        return next.getValue()!!
    }

    override fun iterator(): Iterator<T> {
        return this
    }
}